source/git-api.js

```diff
@@ -31,87 +30,125 @@ exports.registerApi = (env) => {
+        winston.debug(`[${this.watcherId}] path does not exist`, item);
+        return;
+      }
+      const watcher = watch(item, options);
+      watcher.on('change', (_event, changedPath) => {
+        winston.silly(`[${this.watcherId}] ${name}`, changedPath);
+        this.emit(name, changedPath);
+      });
+      this.watchers.push(watcher);
+    }
+    addWorkdir(item, options) {
+      return this.watchItem('workdir', item, options);
+    }
+    addGit(item, options) {
+      return this.watchItem('git', item, options);
+    }
+    close() {
+      this.watchers.forEach((w) => w.close());
+    }
+  }
+
+  const readIgnore = async (pathToWatch) => {
+    winston.debug(`Parsing .gitignore for ${pathToWatch}`);
+    const out = ignore();
+    // For Windows/Mac that can't skip
+    out.add('.git/');
+    const ignoreContent = await fs.readFile(path.join(pathToWatch, '.gitignore')).catch(() => null);
+    if (ignoreContent) out.add(ignoreContent.toString());
+    return out;
+  };
+
+  // TODO move to nodegit
+  const watchRepo = async (pathToWatch, ignore) => {
     winston.info(`Start watching ${pathToWatch}`);
-    return fs
-      .access(pathToWatch)
-      .catch(() => {
-        // Sometimes necessary folders, '.../.git/refs/heads' and etc, are not created on git init
-        winston.debug(`intended folder to watch doesn't exists, creating: ${pathToWatch}`);
-        return mkdirp(pathToWatch);
-      })
-      .then(() => {
-        const watcher = watch(pathToWatch, options || {});
-        watcher.on('change', (event, filename) => {
-          if (!filename) return;
-          const filePath = path.join(subfolderPath, filename);
-          winston.debug(`File change: ${filePath}`);
-          if (isFileWatched(filePath, socket.ignore)) {
-            winston.info(`${filePath} triggered refresh for ${socket.watcherPath}`);
-            emitGitDirectoryChanged(socket.watcherPath);
-            emitWorkingTreeChanged(socket.watcherPath);
+    const watcher = new RepoWatcher();
+    let repoPath = path.join(pathToWatch, '.git');
+    let refsPath;
+    if ((await fs.access(repoPath).catch(() => false)) === undefined) {
+      // Looks like a repo, let's watch workdir
+      let gitIgnore = await readIgnore(pathToWatch);
+      await watcher.addWorkdir(pathToWatch, {
+        recursive: true,
+        filter: (changedPath, skip) => {
+          const filePath = path.relative(pathToWatch, changedPath);
+          if (filePath === '.gitignore') {
+            readIgnore(pathToWatch).then(
+              (ign) => (gitIgnore = ign),
+              (err) => winston.error(`Could not parse .gitignore for`, pathToWatch, err)
+            );
           }
-        });
-        watcher.on('error', (err) => {
-          winston.warn(`Error watching ${pathToWatch}: `, JSON.stringify(err));
-        });
-        socket.watcher.push(watcher);
+          if (filePath === '.git') return skip;
```

Comment:
when it's scanning the workdir and the git repository is in the workdir, it will be encountering `.git`. We're not interested in that.

`.startsWith('.git')` also ignores `.github` which is not correct, so instead I chose to add `.git` to out. But indeed it might be slighty more efficient to add `|| filePath.startsWith('.git'+path.sep)`

Reply:
change applied